api/src/main/java/net/kyori/adventure/text/width/PixelWidthSourceImpl.java

```diff
@@ -0,0 +1,368 @@
+      return this.mergeStylesDownwards(super.render(component, context));
+    }
+
+    @Override
+    protected @Nullable MessageFormat translate(final @NonNull String key, final @NonNull CX context) {
+      return GlobalTranslator.get().translate(key, PixelWidthSourceImpl.this.localeFunction.apply(context));
+    }
+
+    @Override
+    protected @NonNull TextComponent renderBlockNbt(final @NonNull BlockNBTComponent component, final @NonNull CX context) {
+      final BlockNBTComponent component0 = (BlockNBTComponent) super.renderBlockNbt(component, context);
+
+      final TextComponent rendered = this.renderUsingResolver(PixelWidthSourceImpl.this.blockNBTResolvers, component0, context);
+
+      return rendered != null ? rendered : text(component.pos().asString(), component0.style()).children(component0.children());
+    }
+
+    @Override
+    protected @NonNull TextComponent renderEntityNbt(final @NonNull EntityNBTComponent component, final @NonNull CX context) {
+      final EntityNBTComponent component0 = (EntityNBTComponent) super.renderEntityNbt(component, context);
+
+      final TextComponent rendered = this.renderUsingResolver(PixelWidthSourceImpl.this.entityNBTResolvers, component0, context);
+
+      return rendered != null ? rendered : text(component.selector(), component0.style()).children(component0.children());
+    }
+
+    @Override
+    protected @NonNull TextComponent renderStorageNbt(final @NonNull StorageNBTComponent component, final @NonNull CX context) {
+      final StorageNBTComponent component0 = (StorageNBTComponent) super.renderStorageNbt(component, context);
+
+      final TextComponent rendered = this.renderUsingResolver(PixelWidthSourceImpl.this.storageNBTResolvers, component0, context);
+
+      return rendered != null ? rendered : text(component.storage().asString(), component0.style()).children(component0.children());
+    }
+
+    @Override
+    protected @NonNull TextComponent renderKeybind(final @NonNull KeybindComponent component, final @NonNull CX context) {
+      final KeybindComponent component0 = (KeybindComponent) super.renderKeybind(component, context);
+
+      final TextComponent rendered = this.renderUsingResolver(PixelWidthSourceImpl.this.keybindResolvers, component0, context);
+
+      return rendered != null ? rendered : text(component.keybind(), component0.style()).children(component0.children());
+    }
+
+    @Override
+    protected @NonNull TextComponent renderScore(final @NonNull ScoreComponent component, final @NonNull CX context) {
+      final ScoreComponent component0 = (ScoreComponent) super.renderScore(component, context);
+
+      final TextComponent rendered = this.renderUsingResolver(PixelWidthSourceImpl.this.scoreResolvers, component, context);
+
+      return rendered != null ? rendered : text(component.name() + "_" + component.objective() + "_" + component.value(), component0.style()).children(component0.children());
```

Comment:
is this anywhere accurate for a guess?

Reply:
I forgot to mention that in the main post. I was unsure weather to "guess" using info the component has without rendering, or to just ignore it whenever it does not get resolved

Reply:
You should try to match whatever the Vanilla client does for unresolvable content, to give the most accurate results.